### PR TITLE
add unbind

### LINKED ISSUE #2478: unbind API
As suggested in https://github.com/google/flax/pull/2394#discussion_r978744557, its currently not clear for users that `clone` produces an unbounded copy of the module. To get hold of a submodule and its variables you currently have to use this pattern:

```python
module, variables = load_module()
bounded_module = module.bind(variables)
submodule = bounded_module.some_submodule.clone()
submodule_variables = bounded_module.variables
```
The new proposed `.unbind()` method provide more clarity on how to do this process:

```python
module, variables = load_module()
submodule, submodule_variables = module.bind(variables).some_submodule.unbind()
```
Signature could be:
```python
unbind(self, **clone_kwargs) -> Tuple[Module, FrozenVariableDict]
```

cc @jheek

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

Fixes #2478. Adds the `Module.unbind` method which returns an unbound version of the module and its variables.

### Usage
```python
submodule, submodule_vars = module.bind(variables).submodule.unbind()
```

## Context
While `bind` lets the user manipulate the module in a more natural way since it can now rely on direct access fields, there are edge cases and limitations which users needs to be aware of. There is a sentiment that instead of shying away from using `bind` to avoid its pitfalls, we should create documentation around it and add more APIs like `unbind` that make its use a bit more safe in certain cases.

### CODE REVIEW

**Comment by 8bitmp3** [comment]
🔥  Thanks @cgarciae

File: flax/errors.py [hunk 1]
```diff
@@ -569,6 +569,24 @@ def get_submodule(module):
   def __init__(self):
     super().__init__('Can\'t call compact methods on unbound modules')
 
+
+class CallUnbindOnUnboundModuleError(FlaxError):
+  """This error occurs when you are trying to call ``.unbind()`` on an unbound
+  module. For instance, the error will be raised when trying to run this code::
```
Comment (8bitmp3) [inline flax/errors.py:575]:
```suggestion
  """
  This error occurs when you are trying to call ``.unbind()`` on an unbound
  Module. For instance, when you try running the following example,
  an error will be raised:
````

Comment (8bitmp3) [reply flax/errors.py:575]: @cgarciae LMKWYT.

**Comment by andsteing** [comment]
wrt "avoid its pitfalls": maybe it's worth pointing them out in the `.bind()` pydoc?

(from the discussion on #2611 I can mainly remember the potential memory leakage, was there another one to highlight?)

File: flax/linen/module.py [hunk 2]
```diff
@@ -1217,6 +1218,42 @@ def __call__(self, x):
...
 
+  def unbind(self: M) -> Tuple[M, VariableDict]:
+    """Returns an unbound copy of this Module and its variables.
+
+    ``unbind`` is useful for creating a stateless version of a bound Module.
+    A common use case araises when you want extract a submodule defined inside
+    ``setup()``, since ``setup()`` is only called when the Module is bound, a
+    simple way to get hold of the submodule and its corresponding variables is
+    to ``bind`` the parent Module temporarily and then ``unbind`` the desired
+    submodule::
```
Comment (andsteing) [inline flax/linen/module.py:1229]: can / should we add a link to the transfer learning guide?

**Review by andsteing — COMMENTED** [review]
Thanks for rewriting `.extract_field()` to `.bind().unbind()` !

File: flax/linen/module.py [hunk 3]
```diff
@@ -1157,11 +1158,11 @@ def _module_checks(self):
       raise errors.IncorrectPostInitOverrideError()
 
   @traceback_util.api_boundary
-  def bind(self,
+  def bind(self: M,
            variables: VariableDict,
            *args,
            rngs: Optional[RNGSequences] = None,
-           mutable: CollectionFilter = False):
+           mutable: CollectionFilter = False) -> M:
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/module.py:1165]: While `Mbound` could be more descriptive as a type, the current type system cannot express both `M` and `Mbound`.

File: flax/linen/module.py (hunk 2 above)
Comment (cgarciae) [reply flax/linen/module.py:1229]: Will be adding a "Using Stateful Modules guide" soon, we can reference that instead when its ready.

File: docs/guides/transfer_learning.md [hunk 4]
```diff
@@ -72,19 +72,13 @@ Note that `FlaxCLIPVisionModel` itself is not a Flax `Module` which is why we ne
...
-def extract_submodule(clip):
-    vision_model = clip.vision_model.clone()
-    variables = clip.vision_model.variables
-    return vision_model, variables
-
-vision_model, vision_model_variables = nn.apply(extract_submodule, clip)(clip_variables)
+vision_model, vision_model_vars = clip.bind(clip_variables).vision_model.unbind()
 ```
 
 Note that `.clone()` must be used to get an unbounded copy of `vision_model` to avoid leakage as bounded modules contain their variables.
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply docs/guides/transfer_learning.md]: removed

File: flax/linen/module.py [hunk 5]
```diff
@@ -1157,11 +1158,11 @@ def _module_checks(self):
       raise errors.IncorrectPostInitOverrideError()
 
   @traceback_util.api_boundary
-  def bind(self,
+  def bind(self: M,
```
[the remark this answers is not in the document]
Comment (marcvanzee) [reply flax/linen/module.py:1167]: The description "Creates an interactive Module instance by binding variables and RNGs." is a bit odd though, maybe we should just call it a "stateful Module instance", and say it can be used interactivelly?

File: flax/errors.py (hunk 1 above)
Comment (cgarciae) [reply flax/errors.py]: Added.

File: flax/linen/module.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cgarciae) [reply flax/linen/module.py:1264]: @marcvanzee I think we should do this in another PR. Want me to create an separate issue for this?

File: docs/guides/transfer_learning.md [hunk 6]
```diff
@@ -124,13 +118,13 @@ params = variables['params']
```
[the remark this answers is not in the document]
Comment (cgarciae) [reply docs/guides/transfer_learning.md:116]: Done.

**Review by marcvanzee — APPROVED** [review]

File: flax/linen/module.py [hunk 7]
```diff
@@ -1320,6 +1321,42 @@ def __call__(self, x):
...
 
+  def unbind(self: M) -> Tuple[M, VariableDict]:
+    """Returns an unbound copy of this Module and its variables.
+
+    ``unbind`` is useful for creating a stateless version of a bound Module.
+    A common use case araises when you want extract a submodule defined inside
+    ``setup()``, since ``setup()`` is only called when the Module is bound, a
+    simple way to get hold of the submodule and its corresponding variables is
+    to ``bind`` the parent Module temporarily and then ``unbind`` the desired
+    submodule::
```
Comment (8bitmp3) [inline flax/linen/module.py:1332]: Nit:

- sub-Module (we should do this across all docs)
- arises (typo)
- There's a repetition of "You want [to] extract a submodule... to get hold of the submodule...]
- Maybe replace "A common use arises with "An example of a common use case..."
  - Otherwise you have a very complex sentence (in an `unbind` explanation) that may be hard to follow.

Example:

```suggestion
  def unbind(self: M) -> Tuple[M, VariableDict]:
    """Returns an unbound copy of a Module and its variables.

    ``unbind`` helps create a stateless version of a bound Module.

    An example of a common use case: to extract a sub-Module defined inside 
    ``setup()`` and its corresponding variables: 1) temporarily ``bind`` the parent
    Module; and then 2) ``unbind`` the desired sub-Module. (Recall that ``setup()``
    is only called when the Module is bound.)::
```

cc @marcvanzee

File: flax/linen/module.py [hunk 8]
```diff
@@ -1320,6 +1321,42 @@ def __call__(self, x):
...
+          self.encoder = Encoder()
+          self.decoder = Decoder()
+
+        def __call__(self, x):
+          return self.decoder(self.encoder(x))
+
+      module = AutoEncoder()
+      variables = module.init(jax.random.PRNGKey(0), jnp.ones((1, 784)))
+      ...
+      # extract the Encoder submodule and its variables
```
Comment (8bitmp3) [inline flax/linen/module.py:1345]:
```suggestion
      # Extract the Encoder sub-Module and its variables
```

File: flax/errors.py [hunk 9]
```diff
@@ -606,6 +606,30 @@ def get_submodule(module):
   def __init__(self):
     super().__init__('Can\'t call compact methods on unbound modules')
 
+class CallUnbindOnUnboundModuleError(FlaxError):
+  """This error occurs when you are trying to call ``.unbind()`` on an unbound
+  module. For instance, when you try running the following example,
```
Comment (8bitmp3) [inline flax/errors.py:611]:
```suggestion
  Module. For instance, when you try running the following example,
```

File: flax/errors.py [hunk 10]
```diff
@@ -606,6 +606,30 @@ def get_submodule(module):
...
+
+    class MyModule(nn.Module):
+      @nn.compact
+      def __call__(self, x):
+        return nn.Dense(features=10)(x)
+
+    module = MyModule()
+    module.unbind() # <-- ERROR!
+
+  Instead, you should ``bind`` the module to a variable collection before calling
```
Comment (8bitmp3) [inline flax/errors.py:624]:
```suggestion
  Instead, you should ``bind`` the Module to a variable collection before calling
```

File: docs/guides/transfer_learning.md [hunk 11]
```diff
@@ -72,23 +72,15 @@ Note that `FlaxCLIPVisionModel` itself is not a Flax `Module` which is why we ne
 
 Calling `load_model` from the snippet above returns the `FlaxCLIPModule`, which is composed of `text_model` and `vision_model` submodules.
 
-An easy way to extract the `vision_model` submodule defined inside `.setup()` and its variables is to use [`nn.apply`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.apply) to run a `extract_submodule` helper function, inside this function the `clip` Module is bounded to its variables and fields defined in `.setup()` are accessible:
+An easy way to extract the `vision_model` submodule defined inside `.setup()` and its variables is to use [`bind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.bind) on the `clip` Module immediatly followed by [`unbind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.unbind) on the `vision_model` submodule.
```
Comment (8bitmp3) [inline docs/guides/transfer_learning.md:75]: - Using a full method name (`flax.linen.Module.bind`) can help some (new) users identify what is what.
- A minor typo fix.
- "sub-Module".

```suggestion
An easy way to extract the `vision_model` sub-Module defined inside `.setup()` and its variables is to use [`flax.linen.Module.bind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.bind) on the `clip` Module immediately followed by [`flax.linen.Module.unbind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.unbind) on the `vision_model` sub-Module.
```

File: docs/guides/transfer_learning.ipynb [hunk 12]
```diff
@@ -92,7 +92,7 @@
     "\n",
     "Calling `load_model` from the snippet above returns the `FlaxCLIPModule`, which is composed of `text_model` and `vision_model` submodules.\n",
     "\n",
-    "An easy way to extract the `vision_model` submodule defined inside `.setup()` and its variables is to use [`nn.apply`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.apply) to run a `extract_submodule` helper function, inside this function the `clip` Module is bounded to its variables and fields defined in `.setup()` are accessible:"
+    "An easy way to extract the `vision_model` submodule defined inside `.setup()` and its variables is to use [`bind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.bind) on the `clip` Module immediatly followed by [`unbind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.unbind) on the `vision_model` submodule."
```
Comment (8bitmp3) [inline docs/guides/transfer_learning.ipynb:95]: - Using a full method name (`flax.linen.Module.bind`) can help some (new) users identify what is what.
- A minor typo fix.
- "sub-Module".

```suggestion
    "An easy way to extract the `vision_model` sub-Module defined inside `.setup()` and its variables is to use [`flax.linen.Module.bind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.bind) on the `clip` Module immediately followed by [`flax.linen.Module.unbind`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen.html#flax.linen.Module.unbind) on the `vision_model` sub-Module."
```

File: docs/api_reference/flax.linen.rst [hunk 13]
```diff
@@ -91,22 +91,6 @@ Transformations
...
-
-    AxisMetadata
-    map_axis_meta
-    unbox
-    Partitioned
-    with_partitioning
-    get_partition_spec
-
 
 Linear modules
```
Comment (8bitmp3) [inline docs/api_reference/flax.linen.rst:95]:
```suggestion
Linear Modules
```

File: flax/linen/module.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (8bitmp3) [reply flax/linen/module.py:1264]: 👍

**Review by 8bitmp3 — APPROVED** [review]
Minor linting to the unbind docs. LMKWYT @cgarciae

File: flax/linen/module.py [hunk 14]
```diff
@@ -1321,6 +1322,42 @@ def __call__(self, x):
     scope = core.bind(variables, rngs=rngs, mutable=mutable)
     return self.clone(parent=scope)
 
+  def unbind(self: M) -> Tuple[M, VariableDict]:
+    """Returns an unbound copy of a Module and its variables.
+
+    ``unbind`` helps create a stateless version of a bound Module.
+
+    An example of a common use case: to extract a sub-Module defined inside
```
Comment (8bitmp3) [inline flax/linen/module.py:1330]:
```suggestion
    An example of a common use case: to extract a sub-Module defined inside
```
